lib/fluent/plugin/filter_kubernetes_metadata.rb

```diff
@@ -274,35 +250,70 @@ def configure(conf)
 
     end
 
-    def get_metadata_for_record(namespace_name, pod_name, container_name)
+    def get_metadata_for_record(match_data, cache_key, create_time)
+      namespace_name = match_data['namespace']
+      pod_name = match_data['pod_name']
       metadata = {
-        'container_name' => container_name,
+        'container_name' => match_data['container_name'],
         'namespace_name' => namespace_name,
-        'pod_name'       => pod_name,
+        'pod_name'       => pod_name
       }
       if @kubernetes_url.present?
-        cache_key = "#{namespace_name}_#{pod_name}"
-
-        this = self
-        pod_metadata = @cache.getset(cache_key) {
-          @stats.bump(:pod_cache_miss)
-          md = this.get_pod_metadata(
-            namespace_name,
-            pod_name,
-          )
-          md
-        }
+        pod_metadata = get_pod_metadata(cache_key, namespace_name, pod_name, create_time)
         metadata.merge!(pod_metadata) if pod_metadata
+      end
+      metadata
+    end
 
-        if @include_namespace_metadata
-          namespace_metadata = @namespace_cache.getset(namespace_name) {
-            @stats.bump(:namespace_cache_miss)
-            get_namespace_metadata(namespace_name)
+    def get_pod_metadata(key, namespace_name, pod_name, record_create_time)
+        ids = @id_cache.getset(key) do
+          @stats.bump(:id_cache_miss)
+          pod_metadata = fetch_pod_metadata(namespace_name, pod_name)
+          namespace_metadata = fetch_namespace_metadata(namespace_name)
+          ids = {:pod_id=> pod_metadata['pod_id'], :namespace_id => namespace_metadata['namespace_id'] }
+          if ids[:pod_id].nil? && !ids[:namespace_id].nil?
+             ns_time = Time.parse(namespace_metadata['creation_timestamp'])
+             ids[:pod_id] = key if ns_time <= record_create_time #ns is older then record
+          end
+          if ids[:pod_id].nil?
+             @stats.bump(:pod_cache_orphaned_record)
+             ids = :orphaned
+          else
+             @cache[ids[:pod_id]] = pod_metadata
+             @namespace_cache[ids[:namespace_id]] = namespace_metadata
+          end
+          ids
+        end
+        if ids == :orphaned
+          return {} unless @allow_orphans
+          log.trace("orphaning message for: #{namespace_name}/#{pod_name} ") if log.trace?
+          return {
+            'orphaned_namespace' => namespace_name,
+            'namespace_name' => @orphaned_namespace_name,
+            'namespace_id' => @orphaned_namespace_id
           }
-          metadata.merge!(namespace_metadata) if namespace_metadata
         end
-      end
-      metadata
+        metadata =  @cache.getset(ids[:pod_id]) do
```

Comment:
If we have `pod_metadata`, why look this up?  We already know we have set this above.  So we only need to look up in the cache if `pod_metadata` is `nil`, right?  Same is true for `namesapce_metadata` below.

Reply:
Also, we are returning an empty record if we missing looking up the pod ID, but can't we return at least the pod ID as metadata?

Reply:
And we can return the pod name we have in hand, just not as the official pod name, just in another field.

Reply:
This allows us to address the case where metadata is rolled off the cache but the IDs still exist in the ID cache.   Updated to return pod id and namespace id if we still get a miss

Reply:
@jcantrill, I think these comments still hold.  It does not seem to be the case that `getset()` is the right method here.

Reply:
These comments above seem to still hold: it seems we need to avoid double API calls in the case of the `fetch` block being invoked above.

Reply:
In this case, getset is the correct use since it will: fetch and cache and provide a  return value.  The previous us was incorrect as you noted since we returned meta or :orphaned which was the wrong use.